File: backend/app/services/workout_service.py
from typing import List, Optional
from sqlmodel.ext.asyncio.session import AsyncSession
from sqlmodel import select
from datetime import datetime, timedelta
from ..models.workout import WorkoutLog, WorkoutLogCreate, WorkoutType, WorkoutSummary


# MET values (Metabolic Equivalent of Task) for calorie estimation
_MET_VALUES = {
    WorkoutType.CARDIO: 7.0,
    WorkoutType.STRENGTH: 5.0,
    WorkoutType.FLEXIBILITY: 3.0,
    WorkoutType.SPORTS: 6.0,
    WorkoutType.OTHER: 4.0,
}


def estimate_calories(workout_type: WorkoutType, duration_min: int, weight_kg: float) -> int:
    met = _MET_VALUES.get(workout_type, 4.0)
    # Calories = MET * weight_kg * duration_hours
    calories = met * weight_kg * (duration_min / 60.0)
    return round(calories)


class WorkoutService:
    def __init__(self, session: AsyncSession):
        self.session = session

    async def log_workout(self, user_id: int, data: WorkoutLogCreate) -> WorkoutLog:
        workout_data = data.model_dump()
        workout = WorkoutLog(**workout_data, user_id=user_id)
        self.session.add(workout)
        await self.session.commit()
        await self.session.refresh(workout)
        return workout

    async def get_workouts(
        self,
        user_id: int,
        date_from: Optional[datetime] = None,
        date_to: Optional[datetime] = None,
    ) -> List[WorkoutLog]:
        statement = select(WorkoutLog).where(WorkoutLog.user_id == user_id)

        if date_from:
            statement = statement.where(WorkoutLog.created_at >= date_from)
        if date_to:
            statement = statement.where(WorkoutLog.created_at <= date_to)

        statement = statement.order_by(WorkoutLog.created_at.desc())
        result = await self.session.exec(statement)
        return list(result.all())

    async def get_workout_by_id(self, workout_id: int) -> Optional[WorkoutLog]:
        return await self.session.get(WorkoutLog, workout_id)

    async def get_workout_summary(self, user_id: int, days: int = 7) -> WorkoutSummary:
        since = datetime.utcnow() - timedelta(days=days)
        statement = select(WorkoutLog).where(
            WorkoutLog.user_id == user_id,
            WorkoutLog.created_at >= since,
        )
        result = await self.session.exec(statement)
        workouts = list(result.all())

        if not workouts:
            return WorkoutSummary()

        total_duration = sum(w.duration_min for w in workouts)
        total_calories = sum(w.calories_burned or 0 for w in workouts)

        return WorkoutSummary(
            total_workouts=len(workouts),
            total_duration_min=total_duration,
            total_calories=total_calories,
            avg_duration_min=round(total_duration / len(workouts), 1),
        )

    async def delete_workout(self, workout_id: int, user_id: int) -> bool:
        workout = await self.session.get(WorkoutLog, workout_id)
        if not workout or workout.user_id != user_id:
            return False

        await self.session.delete(workout)
        await self.session.commit()
        return True

File: mobile/src/services/workout.service.ts
/**
 * workout.service.ts
 * Workout logging + calorie balance integration with the backend.
 *
 * Network improvements:
 *   - Offline queue for mutations (log, delete)
 *   - In-memory cache for GET responses (workouts, summary, net calories)
 *   - AsyncStorage fallback for offline access to workout data
 *   - Optimistic responses when offline
 *   - Proper TypeScript return types
 *
 * Endpoints:
 *   POST /api/workouts/         -- log a workout (auto-estimates calories via MET)
 *   GET  /api/workouts/         -- list workouts (optional date range)
 *   GET  /api/workouts/summary  -- weekly summary
 *   GET  /api/calories/net      -- net calorie balance for a given day
 *   DELETE /api/workouts/{id}   -- remove a workout
 */
import { api } from './api';
import { getCachedNetworkStatus } from '../hooks/useNetworkStatus';
import { enqueueAction, setCachedData, getCachedData } from './offlineStore';
import {
  cacheResponse,
  getCachedResponse,
  invalidateCache,
  NetworkError,
} from './apiClient';

// ---- Types -------------------------------------------------------------------

export type WorkoutType = 'cardio' | 'strength' | 'flexibility' | 'sports' | 'other';

export interface WorkoutLogCreate {
  workout_type: WorkoutType;
  duration_min: number;
  calories_burned?: number | null;
  notes?: string | null;
}

export interface WorkoutLogRead {
  id: number;
  user_id: number;
  workout_type: WorkoutType;
  duration_min: number;
  calories_burned: number | null;
  notes: string | null;
  created_at: string;
}

export interface WorkoutSummary {
  total_workouts: number;
  total_duration_min: number;
  total_calories: number;
  avg_duration_min: number;
}

export interface ExerciseEntry {
  name: string;
  duration: number;
  calories: number;
  workout_type: string;
}

export interface NetCaloriesResponse {
  date: string;
  consumed: number;
  burned: number;
  net: number;
  goal: number;
  remaining: number;
  deficit_or_surplus: 'deficit' | 'surplus' | 'on_target';
  exercises_today: ExerciseEntry[];
}

// ---- Cache keys --------------------------------------------------------------

const CACHE_KEYS = {
  WORKOUTS: (from?: string, to?: string) => `workouts/list:${from ?? ''}:${to ?? ''}`,
  SUMMARY: (days: number) => `workouts/summary:${days}`,
  NET_CALORIES: (date?: string) => `calories/net:${date ?? 'today'}`,
} as const;

const CACHE_TTL_MS = 5 * 60 * 1000; // 5 minutes

/**
 * Invalidate all workout-related caches after a mutation.
 */
function invalidateWorkoutCaches(): void {
  invalidateCache('workouts/');
  invalidateCache('calories/net');
  // Also invalidate dashboard since workouts affect calorie balance
  invalidateCache('dashboard/today');
}

// ---- Map exercise categories to backend WorkoutType -------------------------

const CATEGORY_TO_WORKOUT_TYPE: Record<string, WorkoutType> = {
  running: 'cardio',
  walking: 'cardio',
  cycling: 'cardio',
  swimming: 'cardio',
  weights: 'strength',
  yoga: 'flexibility',
  hiit: 'cardio',
  sports: 'sports',
};

export function mapCategoryToWorkoutType(category: string): WorkoutType {
  return CATEGORY_TO_WORKOUT_TYPE[category] ?? 'other';
}

// ---- API Calls (offline-aware) -----------------------------------------------

/**
 * Log a workout to the backend.
 * Returns the created workout with auto-estimated calories.
 * Queues the operation when offline and returns an optimistic placeholder.
 */
export const logWorkout = async (data: WorkoutLogCreate): Promise<WorkoutLogRead> => {
  const { isConnected } = getCachedNetworkStatus();
  if (!isConnected) {
    await enqueueAction('log_workout', data);
    invalidateWorkoutCaches();
    // Return optimistic placeholder
    return {
      id: -Date.now(),
      user_id: 0,
      workout_type: data.workout_type,
      duration_min: data.duration_min,
      calories_burned: data.calories_burned ?? null,
      notes: data.notes ?? null,
      created_at: new Date().toISOString(),
    };
  }

  const res = await api.post('/api/workouts/', data);
  invalidateWorkoutCaches();
  return res.data;
};

/**
 * List workouts, optionally filtered by date range.
 * Returns cached data when offline.
 */
export const getWorkouts = async (
  dateFrom?: string,
  dateTo?: string,
): Promise<WorkoutLogRead[]> => {
  const cacheKey = CACHE_KEYS.WORKOUTS(dateFrom, dateTo);

  // Try in-memory cache first
  const cached = getCachedResponse<WorkoutLogRead[]>(cacheKey, CACHE_TTL_MS);
  if (cached) return cached;

  try {
    const params: Record<string, string> = {};
    if (dateFrom) params.date_from = dateFrom;
    if (dateTo) params.date_to = dateTo;
    const res = await api.get('/api/workouts/', { params });
    const workouts: WorkoutLogRead[] = res.data;
    cacheResponse(cacheKey, workouts, CACHE_TTL_MS);
    return workouts;
  } catch (error) {
    if (error instanceof NetworkError && (error.isOffline || error.isTimeout)) {
      // Return empty array when offline with no cache
      return [];
    }
    throw error;
  }
};

/**
 * Get weekly workout summary.
 * Results are cached in memory for 5 minutes.
 */
export const getWorkoutSummary = async (days = 7): Promise<WorkoutSummary> => {
  const cacheKey = CACHE_KEYS.SUMMARY(days);

  const cached = getCachedResponse<WorkoutSummary>(cacheKey, CACHE_TTL_MS);
  if (cached) return cached;

  try {
    const res = await api.get('/api/workouts/summary', { params: { days } });
    const summary: WorkoutSummary = res.data;
    cacheResponse(cacheKey, summary, CACHE_TTL_MS);
    return summary;
  } catch (error) {
    if (error instanceof NetworkError && (error.isOffline || error.isTimeout)) {
      // Return zeroed summary when offline
      return {
        total_workouts: 0,
        total_duration_min: 0,
        total_calories: 0,
        avg_duration_min: 0,
      };
    }
    throw error;
  }
};

/**
 * Get net calorie balance for a given day.
 * Uses in-memory cache (5 min) + offline fallback.
 */
export const getNetCalories = async (date?: string): Promise<NetCaloriesResponse> => {
  const cacheKey = CACHE_KEYS.NET_CALORIES(date);

  const cached = getCachedResponse<NetCaloriesResponse>(cacheKey, CACHE_TTL_MS);
  if (cached) return cached;

  const params: Record<string, string> = {};
  if (date) params.target_date = date;
  const res = await api.get('/api/calories/net', { params });
  const data: NetCaloriesResponse = res.data;
  cacheResponse(cacheKey, data, CACHE_TTL_MS);
  return data;
};

/**
 * Delete a workout by ID.
 * Queues the operation when offline.
 */
export const deleteWorkout = async (workoutId: number): Promise<void> => {
  const { isConnected } = getCachedNetworkStatus();
  if (!isConnected) {
    await enqueueAction('delete_workout', { id: workoutId });
    invalidateWorkoutCaches();
    return;
  }

  await api.delete(`/api/workouts/${workoutId}`);
  invalidateWorkoutCaches();
};
